### PR TITLE
Fixes issue/3496-product-pricing-bug-fix

### LINKED ISSUE #3496: Unable to set 0 price for products
**Describe the bug**

It was reported in 3689003-zen that if a product was created or edited on the WCAndroid app and set the price to 0, the product shows the "Read More" button on the Shop page instead of "Add to Cart".

I was able to reproduce this on WCAndroid but not on WCiOS.

**To Reproduce**
1. Go to Products
2. Create a new product or edit an existing product
3. Enter 0.00 as the price
4. Publish/save the product
5. Visit the Shop page and see that the product has a "Read More" button instead of "Add to Cart"
6. Visiting the product page will not see an option to add the product to the cart as well.

Sample Shop page: https://woo20200908.mystagingwebsite.com/product-category/uncategorized/

Sample product page: https://woo20200908.mystagingwebsite.com/product/free-product/

**Expected behavior**

I would expect the app to allow me to set the product price to 0 so it can be added to the cart as a free product.

Sample product page with $0 price that can be added to the cart: https://woo20200908.mystagingwebsite.com/product/test-free-product/

**Isolating the problem (mark completed items with an [x]):**
- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: OPPO Reno 4
- Android version: 10
- WooCommerce Android version: 5.9-rc-1

**WordPress Environment**
<details>
```
`
### WordPress Environment ###

WordPress address (URL): https://woo20200908.mystagingwebsite.com
Site address (URL): https://woo20200908.mystagingwebsite.com
WC Version: 4.9.2
REST API Version: ✔ 4.9.2
WC Blocks Version: ✔ 4.0.0
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 1.8.3
Log Directory Writable: ✔
WP Version: 5.6
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 7.3.26
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 7.74.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.3.24-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 4.9.2
WC Database Prefix: wp_
Total Database Size: 6.70MB
Database Data Size: 4.72MB
Database Index Size: 1.98MB
wp_woocommerce_sessions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mo2f_network_blocked_ips: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mo2f_network_email_sent_audit: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mo2f_network_transactions: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mo2f_network_whitelisted_ips: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mo2f_user_details: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mo2f_user_login_info: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_options: Data: 0.30MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.22MB + Index: 0.14MB + Engine InnoDB
wp_posts: Data: 0.14MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wfblockediplog: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfblocks7: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wfconfig: Data: 0.47MB + Index: 0.00MB + Engine InnoDB
wp_wfcrawlers: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wffilechanges: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wffilemods: Data: 0.42MB + Index: 0.00MB + Engine InnoDB
wp_wfhits: Data: 0.06MB + Index: 0.05MB + Engine InnoDB
wp_wfhoover: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wfissues: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wfknownfilelist: Data: 1.52MB + Index: 0.00MB + Engine InnoDB
wp_wflivetraffichuman: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wflocs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wflogins: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wfls_2fa_secrets: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wfls_settings: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfnotifications: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfpendingissues: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wfreversecache: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wfsnipcache: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wfstatus: Data: 0.09MB + Index: 0.03MB + Engine InnoDB
wp_wftrafficrates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpns_attack_logs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpns_backup_report: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpns_files_scan: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpns_ip_rate_details: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpns_malware_hash_file: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wpns_malware_scan_report: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpns_malware_scan_report_details: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpns_malware_skip_files: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 56
page: 9
post: 3
product: 32
product_variation: 20
revision: 9
seedprod: 1
shop_coupon: 1
shop_order: 22
shop_order_refund: 2
wc_order_status: 7

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (2) ###

Jetpack by WordPress.com: by Automattic – 9.3.2
WooCommerce: by Automattic – 4.9.2

### Inactive Plugins (13) ###

Akismet Anti-Spam: by Automattic – 4.1.8
All in One Product Quantity for WooCommerce: by WPWhale – 3.6
Coming Soon Page, Maintenance Mode & Landing Pages by SeedProd: by SeedProd – 6.0.10.1
Custom Order Status for WooCommerce: by Tyche Softwares – 2.0.1
Enable jQuery Migrate Helper: by The WordPress Team – 1.3.0
Extra Product Options (Product Addons) for WooCommerce: by ThemeHigh – 3.0.1
miniOrange 2 Factor Authentication: by miniOrange – 5.4.30
Under Construction: by WebFactory Ltd – 3.86
WooCommerce Order Status Manager: by SkyVerge – 1.13.0
WooCommerce Shipment Tracking: by WooCommerce – 1.6.26
WooCommerce Stripe Gateway: by WooCommerce – 4.7.0
Wordfence Security: by Wordfence – 7.4.14
XML-RPC Validator: by Danilo E – 0.1

### Dropin Plugins (2) ###

advanced-cache.php: advanced-cache.php
object-cache.php: Memcached

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #104 - /shop/
Cart: #105 - /cart/
Checkout: #106 - /checkout/
My account: #107 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 3.4.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 23
Oldest: 2020-12-28 09:17:34 +0000
Newest: 2021-01-28 05:37:25 +0000

`
```
</details>

### PR DESCRIPTION
**Pull request by anitaa1990**
Closes: #3496 

## Description
This PR fixes a long standing bug in the Product detail and Variation detail screen where merchants are unable to update a product price to 0. This is possible in Core and in the iOS app. On Android, when the price is set to 0, the product shows the "Read More" button on the Shop page instead of "Add to Cart".

This PR adds logic to treat null or empty price values and 0 price values differently. 
- When the price (regular or sale price) is updated to 0, then the price in the product page should be set to 0.
- When the price (regular or sale price) is updated to an empty or null value, then the price in the product page should not display anything and the `Add to Cart` button should be hidden.
- When the regular price is set to any value and the sale price is set to 0, then the price in the product page should be displayed as 0.
- When the regular price is set to any value and the sale price is set to null, then the price in the product page should be displayed as the regular price.

## Testing instructions

1. Click on the `+` button in the Products tab and create a new product.
2. Notice that the price section displays "Add Price".
3. Click on the "Add Price" section.
4. Click on the back button.
5. Notice the the menu option is not changed in any way.
6. Click on the "Add price" section again and enter any value in the regular price field and sale price field.
7. If the sale price entered is greater than the regular price, then an error message should be displayed.
8. Enter a regular price value greater than 0 and a sale price value less than the regular price and click on the Back button.
9. Notice the new values are updated correctly. Click on the Update/Save button. View the product page and notice the values displayed correctly as entered.
10. Click on the "Add price" section again and update the regular price to 0 and leave the sale price as empty. Save the product. View the product page and notice the price is displayed as 0.
11. Click on the "Add price" section again and leave the regular price empty and leave the sale price as empty. Save the product. View the product page and notice the price is not displayed and the `Add to Cart` button is not displayed.
12. Test the above scenarios for a variation of a product.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/BigDecimalExt.kt [hunk 1]
```diff
@@ -6,13 +6,9 @@ infix fun BigDecimal?.isEqualTo(x: BigDecimal?) = this == x || (x != null && thi
...
 
-infix fun BigDecimal?.isEquivalentTo(that: BigDecimal?): Boolean {
-    val val1 = this ?: BigDecimal.ZERO
-    val val2 = that ?: BigDecimal.ZERO
-    return val1.isEqualTo(val2)
-}
+infix fun BigDecimal?.isEquivalentTo(that: BigDecimal?) = this.isEqualTo(that)
 
-fun BigDecimal?.isNotSet(): Boolean = this.isEquivalentTo(BigDecimal.ZERO)
+fun BigDecimal?.isNotSet(): Boolean = this == null || this.toString().isBlank()
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/BigDecimalExt.kt:11]: @anitaa1990 I _think_ this change is fine but it affects a number of files that aren't modified in this PR. Are we sure this is a safe change?

Screen Shot 2022-03-18 at 7 20 47 AM

**Comment by nbradbury** [comment]
> 5. Notice the Save or Update menu option is NOT displayed.

@anitaa1990 I'm a bit confused by this, because when I tap to create a new product the Publish button is always displayed.

https://user-images.githubusercontent.com/3903757/158995099-d1b529b5-cb0f-4dfe-ba8a-c51eb9131ae7.mov

File: WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/BigDecimalExt.kt (hunk 1 above)
Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/BigDecimalExt.kt:11]: @nbradbury , sorry for not getting back to you sooner! The only places where this change is affected is in the product pricing screen and the product variation pricing screen. And the issue seems to be that, the product pricing and variation pricing cannot be updated to 0 from the app. I double checked that there weren't any other parts of the app using this method so it should be good 👍

**Comment by anitaa1990** [comment]
> I'm a bit confused by this, because when I tap to create a new product the Publish button is always displayed

Ah my bad! I had thought we hide the publish/save button when creating a new product. But the actual testing instructions were to check if there are no changes made to the regular price/sale price, then there should be no update button visible or discard dialog being displayed. 

For a new product specifically, the product detail should display `Add price` and only when the price has been updated to `0` or any other value, that should be reflected in the product detail screen. 

Sorry about the confusion! I'll update the PR description to be more clear

**Review by nbradbury — APPROVED** [review]
Looks good! :shipit:

**Pull request merged**